python/torch_mlir/extras/onnx_importer.py

```diff
@@ -631,6 +758,240 @@ def tensor_proto_to_attr(self, tp: onnx.TensorProto) -> Attribute:
 
 
+def _specialize_function_and_create_model(
+    function_proto: onnx.FunctionProto,
+    op_schema: onnx.defs.OpSchema,
+    name_to_give_model: str,
+    input_type_protos: list[onnx.TypeProto],
+    output_type_protos: list[onnx.TypeProto],
+    caller_node: onnx.NodeProto,
+) -> onnx.ModelProto:
+    """
+    Helper for ModuleCache::get_operator_function() that specializes a function
+    and coverts it to a model.
+
+    An ONNX function may be polymorphic, parameterized over the types of its
+    inputs and values of its attributes (~= compile-time constants). We need to
+    monomorphize it for importing into MLIR. It seems like the only practical
+    way to do this is by turning it into a model:
+    - models can have types on their inputs and outputs, unlike functions
+    - ONNX provides a function to do shape inference (providing concrete
+      types for everything in the body) for models, but not for functions
+    - the rest of the code in this importer can only handle models, not
+      functions
+    """
+
+    graph_proto = onnx.GraphProto()
+
+    for input_name, input_type_proto in zip(function_proto.input, input_type_protos):
+        input_proto = onnx.ValueInfoProto()
+        input_proto.name = input_name
+        input_proto.type.CopyFrom(input_type_proto)
+        graph_proto.input.append(input_proto)
+        output_proto = onnx.ValueInfoProto()
+
+    for output_name, output_type_proto in zip(
+        function_proto.output, output_type_protos
+    ):
+        output_proto.name = output_name
+        output_proto.type.CopyFrom(output_type_proto)
+        graph_proto.output.append(output_proto)
+
+    call_attributes = caller_node.attribute
+    for node in function_proto.node:
+        # Import referenced attributes from call-site or default values
+        new_node = onnx.NodeProto()
+        new_node.CopyFrom(node)
+        old_attributes = list(node.attribute)
+        # .clear() isn't available on protobuf lists for some reason
+        while len(new_node.attribute) > 0:
+            new_node.attribute.pop()
+        for node_attribute in old_attributes:
+            ref_name = node_attribute.ref_attr_name
+            if not ref_name:
+                new_node.attribute.append(node_attribute)
+                continue
+
+            for call_attribute in call_attributes:
+                if call_attribute.name == ref_name:
+                    new_attribute = onnx.AttributeProto()
+                    new_attribute.CopyFrom(call_attribute)
+                    new_attribute.name = node_attribute.name
+                    new_node.attribute.append(new_attribute)
+                    break
+            else:
+                # The default value is sometimes empty for optional attributes
+                # that don't have a default, in which case it is dropped.
+                # cf. https://github.com/onnx/onnx/blob/88f8ef15cfaa3138d336f3502aed5018d802bf43/onnx/shape_inference/attribute_binder.h#L21-L23
```

Comment:
Ah, when I looked again at this reference, I discovered that attribute binding needs to apply to subgraphs! I hadn't considered that. That might explain some of the test failures, I'll need to fix that. (Yet another reason I wish I could use ONNX's existing C++ code for this!)

Reply:
WIP in https://github.com/llvm/torch-mlir/pull/3409/commits/9426f0f6e2424a0d79deae3a4e22c0e30dd9a89d. I hate how this makes the code explosion worse.